src/core/server/http/router/route.ts

```diff
@@ -42,13 +54,68 @@ export interface RouteConfigOptions {
    * Additional metadata tag strings to attach to the route.
    */
   tags?: readonly string[];
+
+  /**
+   * A string or an array of strings with the allowed mime types for the endpoint. Use this settings to limit the set of allowed mime types. Note that allowing additional mime types not listed
+   * above will not enable them to be parsed, and if parse is true, the request will result in an error response.
+   *
+   * Default value: allows parsing of the following mime types:
+   * * application/json
+   * * application/*+json
+   * * application/octet-stream
+   * * application/x-www-form-urlencoded
+   * * multipart/form-data
+   * * text/*
+   */
+  accepts?: Method extends 'get'
+    ? never
+    : RouteContentType | RouteContentType[] | string | string[];
+
+  /**
+   * Limits the size of incoming payloads to the specified byte count. Allowing very large payloads may cause the server to run out of memory.
+   *
+   * Default value: The one set in the kibana.yml config file under the parameter `server.maxPayloadBytes`.
+   */
+  maxBytes?: Method extends 'get' ? never : number;
+
+  /**
+   * The processed payload format. The value must be one of:
+   * * 'data' - the incoming payload is read fully into memory. If parse is true, the payload is parsed (JSON, form-decoded, multipart) based on the 'Content-Type' header. If parse is false, a raw
+   * Buffer is returned.
+   * * 'stream' - the incoming payload is made available via a Stream.Readable interface. If the payload is 'multipart/form-data' and parse is true, field values are presented as text while files
+   * are provided as streams. File streams from a 'multipart/form-data' upload will also have a hapi property containing the filename and headers properties. Note that payload streams for multipart
+   * payloads are a synthetic interface created on top of the entire mutlipart content loaded into memory. To avoid loading large multipart payloads into memory, set parse to false and handle the
+   * multipart payload in the handler using a streaming parser (e.g. pez).
+   * * 'file' - the incoming payload is written to temporary file in the directory specified by the uploads settings. If the payload is 'multipart/form-data' and parse is true, field values are
+   * presented as text while files are saved to disk. Note that it is the sole responsibility of the application to clean up the files generated by the framework. This can be done by keeping track
+   * of which files are used (e.g. using the request.app object), and listening to the server 'response' event to perform cleanup.
+   *
+   * Default value: 'data'.
+   */
+  output?: Method extends 'get' ? never : 'data' | 'stream' | 'file';
```

Comment:
__question:__ These questions are just for my own understanding, please feel free to ignore (don't want to block as I'm OOO anyway :slightly_smiling_face: ):

* `.. specified by the uploads settings` --- > Do we use `file` anywhere in Kibana already? If not, maybe it'd make sense to not expose this for now to keep exposed API surface manageable and less dependent on Hapi? True for all other options we're exposing here.
* I always felt that some of the Hapi names are really confusing, can we use our own more descriptive names since it's a brand new API for core anyway (e.g. `output` ---> `payloadFormat`, `bodyFormat` or something similar)?
* I feel a bit worried that `parse` isn't tied to `validate` in any way. I believe you had discussion around that already and I'm just missing context, but can't we set `parse` to `false` if `validate` is set to `false` for example or at least set better/safer defaults (e.g. `parse: false` and `output: 'stream'`).
* Do we validate at run-time that all these parameters corresponds to what we require via types or we solely rely on Hapi to validate those? If the latter we may be in troubles in the future when we upgrade Hapi and it introduces new options we don't want to support or just handle differently. 

As you see I'm just alluding to the fact that we may not want to copy Hapi APIs and instead model our own public API with better defaults that would rely only on certain parts of the Hapi APIs under the hood. /cc @restrry @kobelb

Reply:
Hi @azasypkin, thank you for your comments, I really appreciate them.

> `.. specified by the uploads settings` --- > Do we use `file` anywhere in Kibana already? If not, maybe it'd make sense to not expose this for now to keep exposed API surface manageable and less dependent on Hapi? True for all other options we're exposing here.

I agree. Since it is not currently used, maybe it's better to hide it from the docs (and use a validation step to reject it if provided despite types?)

> I always felt that some of the Hapi names are really confusing, can we use our own more descriptive names since it's a brand new API for core anyway (e.g. `output` ---> `payloadFormat`, `bodyFormat` or something similar)?

This is kind of my fault, the actual property from HAPI is `payload.output`. Maybe it provides more context. I brought these 4 properties up to the top-level options thinking this way we don't expose the raw `payload` option to the user (where they could do `{anyOtherNotExposedOption} as any` to workaround us). But I can group them down to `payload` and still write the code in a defensive way.

> I feel a bit worried that `parse` isn't tied to `validate` in any way. I believe you had discussion around that already and I'm just missing context, but can't we set `parse` to `false` if `validate` is set to `false` for example or at least set better/safer defaults (e.g. `parse: false` and `output: 'stream'`).

I like the idea of setting `parse: false` and `output: 'stream'` if `validate === false`. That should release some memory pressure. I wonder what we should do when `validate` exists but doesn't contain the `body` property?

> Do we validate at run-time that all these parameters corresponds to what we require via types or we solely rely on Hapi to validate those? If the latter we may be in troubles in the future when we upgrade Hapi and it introduces new options we don't want to support or just handle differently.

Hapi does some validation under the hood for "sensible" values, but you are right. We should be more defensive regarding our accepted values (over the wider options provided by Hapi). I'll see what I can do with this.

Reply:
Updated, taking into account this conversation:
1. Moved the payload-specific config inside a property called `body` to match the same naming convention we've got in `validate`
2. Added validation to avoid `body.output` to be specified as anything different to 'data' or 'stream'.
3. When `validate.body` does not exist, we assume the route does not care about the body content, so we default to `output: 'stream'` and `parse: false`